implement more network apis

d3m3vilurr opened the pull request:
- select of sys/select.h
- gethostbyname of netdb.h
- add macros of netinet/in.h
- add sys/types.h for custom size fd_set
- reduce MAX_OPEN_FILES to 256

d3m3vilurr committed:
add more network api & macro

- select of sys/select.h
- gethostbyname of netdb.h
- add macros of netinet/in.h

xyzz commented on newlib/libc/sys/vita/select.c:
```diff
@@ -0,0 +1,128 @@
+				inc = 1;
+				FD_SET(events[i].data.fd, exceptfds);
+			}
+            res += inc;
```
is this correct behavior? Or should it instead increment once per event?

http://pubs.opengroup.org/onlinepubs/9699919799/functions/select.html says

> Upon successful completion, the pselect() and select() functions shall return the total number of bits set in the bit masks. Otherwise, -1 shall be returned, and errno shall be set to indicate the error.

so I feel like the original behavior was correct here

d3m3vilurr replied:
your original code use if-else blocks. so some events can be ignore.
i think, res value should be increase each FD_SET.. i will change that.

xyzz commented on newlib/libc/sys/vita/sys/socket.h:
```diff
@@ -247,6 +262,24 @@ struct msghdr {
 
+#define HAVE_SYS_SOCKET_H
+
+#define HAVE_RECV
```
these defines below seem to be curl-specific, why are you adding them to newlib?

d3m3vilurr replied:
huh...removed. i think these macros are standard :(

xyzz commented on newlib/libc/sys/vita/select.c:
```diff
@@ -0,0 +1,128 @@
+		__vita_fd_drop(fdmap);									\
+	} while(0)
+
+int select(int nfds, fd_set *readfds, fd_set *writefds, fd_set *exceptfds,
```
this function should be thread-safe. not sure if it's worth adding thread-safety here though, ideally we'd grab all required FDs before creating the epoll, and drop them after wait returns.

d3m3vilurr replied:
actually method outside use wrapped fd value. but inside, use `sce_uid` variable.
so i use these action instead `__vita_fdmap[fd]`. want you to change this?

xyzz replied:
yes I understand why it's used like that. but you grab a fd and then drop it immediately, so it's not thread-safe this way - somebody can close() a socket while select() is operating

d3m3vilurr replied:
hm....ok. we need manual mutex handling right?

DaveeFTW replied:
You need to handle other file descriptor types, too. This implementation is strictly for socket types.

jdek commented:
Squash these commits.

d3m3vilurr commented:
i will squash after review :)

xyzz commented on newlib/libc/sys/vita/unistd.c:
```diff
@@ -0,0 +1,38 @@
+#include <errno.h>
+#include "vitadescriptor.h"
+
+int close(int s)
```
there's already a close() implementation available in `syscalls.c`

d3m3vilurr replied:
huh... ok i will remove these..

DaveeFTW commented on newlib/libc/sys/vita/sys/socket.h:
```diff
@@ -160,6 +162,16 @@ struct	linger {
+#define IPPROTO_ICMP    SCE_NET_IPPROTO_ICMP
+#define IPPROTO_IGMP    SCE_NET_IPPROTO_IGMP
+#define IPPROTO_TCP     SCE_NET_IPPROTO_TCP
+#define IPPROTO_UDP     SCE_NET_IPPROTO_UDP
```
These macros don't belong in sys/socket.h

d3m3vilurr replied:
i will mode to netinet/in.h

DaveeFTW commented on newlib/libc/sys/vita/sys/socket.h:
```diff
@@ -169,6 +181,9 @@ struct sockaddr {
 
+#define in_addr     SceNetInAddr
+#define sockaddr_in SceNetSockaddrIn
+
```
These structures don't belong in sys/socket.h

DaveeFTW commented on newlib/libc/sys/vita/sys/socket.h:
```diff
@@ -267,6 +282,11 @@ int	setsockopt(int, int, int, const void *, socklen_t);
+#define ntohs __builtin_bswap16
+#define htons __builtin_bswap16
+#define ntohl __builtin_bswap32
+#define htonl __builtin_bswap32
```
These don't belong in sys/socket either.

DaveeFTW commented on newlib/libc/sys/vita/sys/select.h:
```diff
@@ -0,0 +1,49 @@
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef _SYS_SOCKET_H_
+#define _SYS_SOCKET_H_
```
Change include guards.

d3m3vilurr commented:
fixed all review comments & suggest in irc
1. squashed commits
2. define values move right position.
3. remove wrong edits; wrong file, unused functions...
4. add locked array. this uint8_t array use 128bytes(1028bits) for marking grabbed fd idx, then reuse for drop.

xyzz commented on newlib/libc/sys/vita/select.c:
```diff
@@ -0,0 +1,143 @@
+
+#define LOCK_ARRAY_LENGTH   (MAX_OPEN_FILES / 8) + ((MAX_OPEN_FILES % 8) ? 1 : 0)
+#define BIT_SET(arr, idx)	(arr)[(idx)/8] |= (1<<((idx)%8))
+#define IS_SET(arr, idx)	((arr)[(idx)/8] & (1<<((idx)%8))) != 0
```
you probably can just reuse `FD_SET` family of functions

d3m3vilurr replied:
i think cannot direct use FD_SET.
FD_SET limitation is these macros only allow fd_set type.
and newlibs defined FD_SETSIZE is 64 not 128.

I think we use too many MAX_OPEN_FILES value...

xyzz commented on newlib/libc/sys/vita/select.c:
```diff
@@ -0,0 +1,143 @@
+		return 0;
+	}
+
+	uint8_t locked[LOCK_ARRAY_LENGTH];
```
`uint8_t locked[LOCK_ARRAY_LENGTH] = {0}`

d3m3vilurr replied:
fixed

xyzz commented on newlib/libc/sys/vita/select.c:
```diff
@@ -0,0 +1,143 @@
+	sceNetEpollDestroy(eid);
+exit:
+	for (i = 0; i < nfds; i++) {
+		if (IS_SET(locked, i) && __vita_fdmap[i] != NULL) {
```
I'd probably use `__vita_fd_grab` here instead and then `__vita_fd_drop` it twice, need a comment from @DaveeFTW though

d3m3vilurr commented:
we need reduce MAX_OPEN_FILES to 64 or define FD_SETSIZE to 1024 for this feature.
(or custom define fd_set type like linux https://github.com/vitasdk/newlib/blob/vita/newlib/libc/sys/linux/sys/types.h#L119)
newlib's default FD_SETSIZE of `sys/types.h` use 64, it is not fit about vitadescriptor.h's define.
and i'm not tested yet, but vita may not support 1024 fds. (it's too many fd for embedded device)

currently i don't know, this is right or not. so someone confirm this, i will append patch for this.

d3m3vilurr committed:
Reduce MAX_OPEN_FILES to 256

d3m3vilurr committed:
Add custom sys/types.h

This file have custom fd_set; this struct contain 256bits
Now locked list can use fd_set instead of uint8_t array.

DaveeFTW commented on newlib/libc/sys/vita/select.c:
```diff
@@ -0,0 +1,138 @@
+			continue;
+		}
+
+		DescriptorTranslation *fdmap = __vita_fd_grab(i);
```
You need to retain this value in order to ensure you are dropping the correct reference. Otherwise race conditions where the descriptor is destroyed may break prevent the resource from being released.

d3m3vilurr replied:
hm...
this patch use `lock` variable for store grabbed fd numbers.
so still can race condition?

DaveeFTW replied:
It can race condition between this point and the comment left here: https://github.com/vitasdk/newlib/pull/9/files#r80349878

The problem lies under the use of `__vita_fdmap`, which may no longer point to the correct fd. You must drop the pointer returned from `__vita_fd_grab`.

d3m3vilurr replied:
changed `fd_set lock = {0}` to `DescriptorTranslation *lock[MAX_OPEN_FILES] = {0}`
now direct use dropped pointer. (but lock memory increase to 1K..)

if i have right understand, someone grap fd, this ref never release before drop.
so still __vita_fdmap[i] == returned grap pointer
but it depends implements, so `drop` implement can change something, we still cannot use pointer as same reason.

how about regrap then double drop like xyz said?
(so now, i'm understand why he said that.)

maybe like this..

```
if (lock[i]) {
  tmp = __vita_fd_grab(i);
  if (tmp.sce_uid == lock[i].sce_uid && tmp.type == lock[i].type) {
    // something changed...
    __vita_fd_drop(lock[i]);
  }
  __vita_fd_drop(tmp);
}
```

DaveeFTW commented on newlib/libc/sys/vita/select.c:
```diff
@@ -0,0 +1,138 @@
+exit:
+	for (i = 0; i < nfds; i++) {
+		if (FD_ISSET(i, &lock) && __vita_fdmap[i] != NULL) {
+			__vita_fd_drop(__vita_fdmap[i]);
```
As above. If fd X is deleted immediately before this loop then the underlying block will not drop reference counter properly and clean up the resource.

>   Multithreaded applications
>        If a file descriptor being monitored by select() is closed in another
>        thread, the result is unspecified.  On some UNIX systems, select()
>        unblocks and returns, with an indication that the file descriptor is
>        ready (a subsequent I/O operation will likely fail with an error,
>        unless another the file descriptor reopened between the time select()
>        returned and the I/O operations was performed).  On Linux (and some
>        other systems), closing the file descriptor in another thread has no
>        effect on select().  In summary, any application that relies on a
>        particular behavior in this scenario must be considered buggy.

http://man7.org/linux/man-pages/man2/select.2.html
In our case, if close is called within select it the fd_grab will protect the underlying resource.

DaveeFTW commented on newlib/libc/sys/vita/select.c:
```diff
@@ -0,0 +1,138 @@
+
+	if (nfds == 0) {
+		sceKernelDelayThread(wait);
+		return 0;
```
errno is not set here.

d3m3vilurr replied:
i think, it's normal case.
just guard code for avoid useless epoll call. so just wait, then return `normal`
i guess, we should change to `if (nfds <= 2) {` (avoid stdin~stderr) but don't know right or not.

DaveeFTW replied:
This code is correct, in fact. It seems its non-standard (maybe even against standard) to store 0 in errno within libraries.

d3m3vilurr replied:
ok i added `errno = 0`

DaveeFTW replied:
I made a mistake, I am trying to say that your code correct and you do _not_ need to set errno to 0.

d3m3vilurr replied:
reverted. :) btw, it's my problem about english skill.. sorry.

DaveeFTW reviewed (changes requested):
I imagine if you have to store a list of grabbed fd references it would be fairly straightforward to implement both select() and poll() using an abstracted implementation.

d3m3vilurr committed:
Fix can race condition code

Direct use DescriptorTranslation* array instead bitflags

DaveeFTW reviewed (changes requested):
You must check that `nfds` parameter is less than `MAX_OPEN_FILES` and set `EINVAL` appropriately.

d3m3vilurr committed:
`nfds` should be 0 to MAX_OPEN_FILES - 1

d3m3vilurr commented:
Add guard code for `ndfs` range.
